blockchain/Validation.go

```diff
@@ -1490,7 +1491,132 @@ func FindFilesToPrune(setFilesToPrune *set.Set, nPruneAfterHeight uint64) {
-	//TODO
+	var (
+		m      *mempool.Mempool
+		params *msg.BitcoinParams
+	)
+	mempoolUsage := m.DynamicMemoryUsage()
+	//TODO: LOCK2(cs_main, cs_LastBlockFile);
+	//var sc sync.RWMutex
+	//sc.Lock()
+	//defer sc.Unlock()
+	nLastWrite := 0
+	nLastFlush := 0
+	nLastSetChain := 0
+
+	var setFilesToPrune *set.Set
+	fFlushForPrune := false
+
+	if r := recover(); r != nil {
+		if GfPruneMode && (GfCheckForPruning || nManualPruneHeight > 0) && !GfReindex {
+			FindFilesToPruneManual(setFilesToPrune, nManualPruneHeight)
+		} else {
+			FindFilesToPrune(setFilesToPrune, uint64(params.PruneAfterHeight))
+			GfCheckForPruning = false
+		}
+		if !setFilesToPrune.IsEmpty() {
+			fFlushForPrune = true
+			if !GfHavePruned {
+				//TODO: pblocktree.WriteFlag("prunedblockfiles", true)
+				GfHavePruned = true
+			}
+		}
+		nNow := utils.GetMockTimeInMicros()
+		// Avoid writing/flushing immediately after startup.
+		if nLastWrite == 0 {
+			nLastWrite = int(nNow)
+		}
+		if nLastFlush == 0 {
+			nLastFlush = int(nNow)
+		}
+		if nLastSetChain == 0 {
+			nLastSetChain = int(nNow)
+		}
+
+		nMempoolSizeMax := utils.GetArg("-maxmempool", int64(policy.DEFAULT_MAX_MEMPOOL_SIZE)) * 1000000
+		cacheSize := GpcoinsTip.DynamicMemoryUsage() * DB_PEAK_USAGE_FACTOR
+		nTotalSpace := float64(GnCoinCacheUsage) + math.Max(float64(nMempoolSizeMax-mempoolUsage), 0)
+		// The cache is large and we're within 10% and 200 MiB or 50% and 50MiB
+		// of the limit, but we have time now (not in the middle of a block processing).
+		x := math.Max(nTotalSpace/2, nTotalSpace-MIN_BLOCK_COINSDB_USAGE*1024*1024)
+		y := math.Max(9*nTotalSpace/10, nTotalSpace-MAX_BLOCK_COINSDB_USAGE*1024*1024)
+		fCacheLarge := mode == FLUSH_STATE_PERIODIC && float64(cacheSize) > math.Min(x, y)
+		// The cache is over the limit, we have to write now.
+		fCacheCritical := mode == FLUSH_STATE_IF_NEEDED && float64(cacheSize) > nTotalSpace
+		// It's been a while since we wrote the block index to disk. Do this
+		// frequently, so we don't need to redownload after a crash.
+		fPeriodicWrite := mode == FLUSH_STATE_PERIODIC && int(nNow) > nLastWrite+DATABASE_WRITE_INTERVAL*1000000
+		// It's been very long since we flushed the cache. Do this infrequently,
+		// to optimize cache usage.
+		fPeriodicFlush := mode == FLUSH_STATE_PERIODIC && int(nNow) > nLastFlush+DATABASE_FLUSH_INTERVAL*1000000
+		// Combine all conditions that result in a full cache flush.
+		fDoFullFlush := mode == FLUSH_STATE_ALWAYS || fCacheLarge || fCacheCritical || fPeriodicFlush || fFlushForPrune
+		// Write blocks and block index to disk.
+		if fDoFullFlush || fPeriodicWrite {
+			// Depend on nMinDiskSpace to ensure we can write block index
+			if !CheckDiskSpace(0) {
+				return state.Error("out of disk space")
+			}
+			// First make sure all block and undo data is flushed to disk.
+			FlushBlockFile(false)
+			// Then update all block file information (which may refer to block and undo files).
+
+			var files map[int]*BlockFileInfo
```

Comment:
the origin type  is vector in C++, so the vector container has order by user; when you replace with map, the order will modify by the underlying mechanism of map. so when you write these data into file, the store data is different with origin file.